### PR TITLE
Feature some additional collectives

### PR DESCRIPTION
**Pull request by benson31**
This adds MPI-CUDA backend support for Alltoall, Scatter, and Gather. The implementation is similar to HostTransferState for Allreduce, but it uses host-based nonblocking MPI to do the communication. This allows us to enqueue all collectives in Hydrogen on streams.

Combined with #3, tests with Alexnet on pascal now show consistent and better performance.

### CODE REVIEW

File: src/mpi_cuda/gather.hpp [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+        else
+          return true;
+      }
+      else {
+        return false;
+      }
+    }
+    else if (rank_ != root_) {
+      // Paranoia, in case step() is ever called again after returning
+      // 'true' for the first time.
```
Comment (ndryden) [inline src/mpi_cuda/gather.hpp]: If that happens, I'd call it a bug in the progress engine, but fair enough.

File: src/mpi_cuda/gather.hpp [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+      else if (r == cudaErrorNotReady) {
+        return false;
+      }
+      else {
+        throw_al_exception("Alltoall: cudaEventQuery error");
+      }
+    }
+
+    if (!gather_done_) {
+      // Wait for the all2all to complete
```
Comment (ndryden) [inline src/mpi_cuda/gather.hpp]: gather

File: src/mpi_cuda/scatter.hpp [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+class ScatterAlState : public AlState {
+public:
+  ScatterAlState(const T* sendbuf, T* recvbuf, size_t count, int root,
+                MPICUDACommunicator& comm, cudaStream_t stream) :
+    AlState(nullptr),
+    rank_(comm.rank()), root_(root), count_(count),
+    host_mem_(get_pinned_memory<T>(rank_ == root_
+                                   ? comm.size()*count_ : count_)),
+    d2h_event_(rank_ == root_ ? cuda::get_cuda_event() : nullptr),
+    h2d_event_(cuda::get_cuda_event()),
```
Comment (ndryden) [inline src/mpi_cuda/scatter.hpp]: Ditto.

File: src/mpi_cuda/scatter.hpp [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+      else if (r == cudaErrorNotReady) {
+        return false;
+      }
+      else {
+        throw_al_exception("Alltoall: cudaEventQuery error");
+      }
+    }
+
+    if (!scatter_done_) {
+      // Wait for the all2all to complete
```
Comment (ndryden) [inline src/mpi_cuda/scatter.hpp]: scatter

File: src/Al.hpp [hunk 5]
```diff
@@ -442,6 +442,186 @@ void NonblockingBcast(
...
+void Gather(
+  const T* sendbuf, T* recvbuf, size_t count, int root,
+  typename Backend::comm_type& comm,
+  typename Backend::algo_type algo = Backend::algo_type::automatic) {
+  Backend::template Gather<T>(sendbuf, recvbuf, count, root, comm, algo);
+}
+
+/**
+ * In-place gather-to-one operation.
+
```
[the remark this answers is not in the document]
Comment (benson31) [reply src/Al.hpp]: Fixed

File: src/mpi_cuda/alltoall.hpp [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+
+template <typename T>
+class AlltoallAlState : public AlState {
+public:
+  AlltoallAlState(const T* sendbuf, T* recvbuf, size_t count,
+                  MPICUDACommunicator& comm, cudaStream_t stream) :
+    AlState(nullptr),
+    host_mem_(get_pinned_memory<T>(comm.size()*count)),
+    count_(count),
+    d2h_event_(cuda::get_cuda_event()),
```
[the remark this answers is not in the document]
Comment (benson31) [reply src/mpi_cuda/alltoall.hpp]: Changed to FastEvent. ON systems with no stream mem ops, this is equivalent after #3 merges.

File: src/Al.hpp [hunk 7]
```diff
@@ -442,6 +442,186 @@ void NonblockingBcast(
...
+void Scatter(
+  const T* sendbuf, T* recvbuf, size_t count, int root,
+  typename Backend::comm_type& comm,
+  typename Backend::algo_type algo = Backend::algo_type::automatic) {
+  Backend::template Scatter<T>(sendbuf, recvbuf, count, root, comm, algo);
+}
+
+/**
+ * In-place scatter-to-all operation.
+
```
[the remark this answers is not in the document]
Comment (benson31) [reply src/Al.hpp]: Fixed

File: src/mpi_cuda/gather.hpp (hunk 2 above)
Comment (benson31) [reply src/mpi_cuda/gather.hpp]: Fixed all.

File: src/mpi_cuda/gather.hpp (hunk 1 above)
Comment (benson31) [reply src/mpi_cuda/gather.hpp]: I agree, but it's a simple enough check...

**Pull request merged**